### PR TITLE
Make compatible with PG15

### PR DESCRIPTION
**Pull request by keithf4**
Make compatible with PG15. The following upstream PRs renamed `scanint8()` to a new function and removed the error checking method that pgnodemx had been using before. Change to using `strtoi64()` at suggestion of first PR, which is also a new function in 15.

https://github.com/postgres/postgres/commit/cfc7191dfea330dd7a71e940d59de78129bb6175
https://github.com/postgres/postgres/commit/3c6f8c011f85df7b35c32f4ccaac5c86c9064a4a

Review Request Note:

Not sure the way I'm handling the error condition with `errno` is the same as it was doing before. Either need confirmation from someone else with more knowledge on its usage with the str -> int conversion functions or actually testing the error condition with a K8 setup.

Review Update:

Jan provided some of the requested review above. Thank you!

### CODE REVIEW

**Review by youattd — APPROVED** [review]

File: cgroup.c [hunk 1]
```diff
@@ -150,8 +151,21 @@ cgmembers(int64 **pids)
...
 		bool	success = false;
 		int64	result;
+		#if PG_VERSION_NUM >= 150000
+		char   *endptr;
+		#endif
 
+		#if PG_VERSION_NUM < 150000
 		success = scanint8(lines[i], true, &result);
+		#endif
+		#if PG_VERSION_NUM >= 150000
```
Comment (pgguru) [inline cgroup.c]: Stylistically, I'd probably do a `#else` here, unless there's precedence for doing it this way in the other code.

File: parseutils.c [hunk 2]
```diff
@@ -560,7 +565,17 @@ get_int64_from_file(char *ftr)
 		result = PG_INT64_MAX;
 	else
 	{
+        #if PG_VERSION_NUM < 150000
```
Comment (pgguru) [inline parseutils.c]: I'd unindent all of these inline preprocessor macros; at least that's how I've seen this sort of version-specific code. (Again, project practice here.)  As currently written, it's a little hard to tease out the logic flow.

File: parseutils.c [hunk 3]
```diff
@@ -560,7 +565,17 @@ get_int64_from_file(char *ftr)
...
 		success = scanint8(rawstr, true, &result);
+		#endif
+		#if PG_VERSION_NUM >= 150000
+		errno = 0;
+		result = strtoi64(rawstr, &endptr, 10);
+		if (errno == 0)
+		{
+			success = true;
+		}
+		#endif
```
Comment (pgguru) [inline parseutils.c]: I'd probably do something like:
```suggestion
#else
		errno = 0;
		result = strtoi64(rawstr, &endptr, 10);
		success = !errno;
#endif
```

File: pgnodemx.c [hunk 4]
```diff
@@ -596,11 +598,24 @@ pgnodemx_envvar_bigint(PG_FUNCTION_ARGS)
...
+	char   *endptr;
+	#endif
 
 	/* Limit use to members of special role */
 	pgnodemx_check_role();
 
+	#if PG_VERSION_NUM < 150000
 	success = scanint8(value, true, &result);
+	#endif
+	#if PG_VERSION_NUM >= 150000
```
Comment (pgguru) [inline pgnodemx.c]: Same comments apply to this one.

**Review by pgguru — CHANGES_REQUESTED** [review]

File: cgroup.c (hunk 1 above)
Comment (keithf4) [reply cgroup.c]: Done

File: parseutils.c (hunk 2 above)
Comment (keithf4) [reply parseutils.c]: Didn't even notice that for the other one. Thanks. Fixed for consistency.

File: parseutils.c (hunk 3 above)
Comment (keithf4) [reply parseutils.c]: Done

File: pgnodemx.c (hunk 4 above)
Comment (keithf4) [reply pgnodemx.c]: Done

**Comment by youattd** [comment]
Agree on moving the c-preprocessor macros so the `#` starts in the first column.  The cpp used to require that.
Also note that there is a `#elif` in cpp if needed.
Most `errno`s are set by system calls, a few by libc calls.  Reviewing the changes now in that context.

**Comment by youattd** [comment]
For `errno` handling, it looks like pg 15's `strtoi64` is a macro for C lib `strtol` (and `strtoll`).  Check the man page for `strtol` but I think it sets `errno` for integer overflow, so the `errno` handling should be ok.

Also, if it fits and helps simplify, the C-preprocessor also has  `#if defined ...` where you can combine conditions using `&&` and `||` in to expressions that are evaluated by the `defined`

**Review by wieck — APPROVED** [review]
Changes look proper to me.

**Comment by keithf4** [comment]
> For `errno` handling, it looks like pg 15's `strtoi64` is a macro for C lib `strtol` (and `strtoll`). Check the man page for `strtol` but I think it sets `errno` for integer overflow, so the `errno` handling should be ok.
> 
> Also, if it fits and helps simplify, the C-preprocessor also has `#if defined ...` where you can combine conditions using `&&` and `||` in to expressions that are evaluated by the `defined`

@youattd  Thanks I did look at the `strtol` reference and saw that it doesn't always set the errno for some "invalid" situations, hence why the source upstream PR referenced code has an example that also checks the `endptr` to actually be the end of the value. Added that back in to my check as well after consulting with Jan. Let me know how that looks.

**Review by pgguru — APPROVED** [review]
As long as you've tested against pre-15, I'd assume this is fine; my one question now is whether moving the location of the `#include utils/int8.h` relative to where it was causes any issues.  Would seem unlikely, which is why I'm approving, just something to consider.

**Review by youattd — APPROVED** [review]

**Review by wieck — APPROVED** [review]

**Review by andrewlecuyer — APPROVED** [review]

**Pull request merged**